### PR TITLE
[#3881] Update variables in reusable form definitions

### LINKED ISSUE #3881: Two forms with a re-usable form definition "sometimes" leads to an incorrect form in one of the two
We found this issue when figuring out why voorletters was suddenly not working in the prefill when migrating from 2.4.3 to 2.4.5.

The client said it didn't work. Made the initials field writable (it was readonly) so it could be filled, and saved the form (this step was a re-usable form definition). When we tested it, it worked. Silvia suggested it could be related to a variable not being created and it was then created when the form was saved, making the form function.

This might also be the case why a re-usable form definition works in form A but not in form B, after the form definition is changed in form A.

### PR DESCRIPTION
**Pull request by vaszig**
Fixes #3881

### CODE REVIEW

File: src/openforms/forms/tasks.py [hunk 1]
```diff
@@ -14,6 +15,23 @@
 logger = logging.getLogger(__name__)
 
 
+def on_variables_bulk_update_event(form_id: int) -> None:
+    """
+    Celery chain of tasks to execute on a bulk update of variables completion event.
```
Comment (sergei-maertens) [inline src/openforms/forms/tasks.py:20]:
```suggestion
    Celery chain of tasks to execute on a bulk update of variables event.
```

File: src/openforms/forms/tasks.py [hunk 2]
```diff
@@ -14,6 +15,23 @@
...
+    """
+    recouple_reusable_definition_variables_task = (
+        recouple_reusable_definition_variables_to_form_variables.si(form_id)
+    )
```
Comment (sergei-maertens) [inline src/openforms/forms/tasks.py:24]:
```suggestion
    repopulate_reusable_definition_variables_task = (
        recouple_reusable_definition_variables_to_form_variables.si(form_id)
    )
```
perhaps? it's not necessarily fixing broken relations, rather it's deriving the variables from the new FD again :thinking:

File: src/openforms/forms/tasks.py [hunk 3]
```diff
@@ -87,6 +105,33 @@ def recouple_submission_variables_to_form_variables(form_id: int) -> None:
...
 
 
+@app.task(ignore_result=True)
+def recouple_reusable_definition_variables_to_form_variables(form_id: int) -> None:
+    """
+    When the FormVariable bulk create/update endpoint is called, all existing FormVariable related to the form are
+    deleted and new are created. If there are any form definitions which are reusable, we want to update all the forms
+    which are also using these FormDefinitions (concerning the FormVariables). This task updates the FormVariables
+    of each related form in the database, by deleteing the old ones and recreating them (the most safe way).
+    """
```
Comment (sergei-maertens) [inline src/openforms/forms/tasks.py:115]:
```suggestion
    """
    Fix inconsistencies created by updating a re-usable form definition which is used in other forms.

    When the FormVariable bulk create/update endpoint is called, all existing FormVariable related to the form are
    deleted and new are created. If there are any form definitions which are reusable, we want to update all the forms
    which are also using these FormDefinitions (concerning the FormVariables). This task updates the FormVariables
    of each related form in the database, by replacing the old state with the newly derived set of form variables.
    """
```
in docstrings, you should always provide a one-line summary and the put the extra information below it. See PEP-257 :)

File: src/openforms/forms/tasks.py [hunk 4]
```diff
@@ -87,6 +105,33 @@ def recouple_submission_variables_to_form_variables(form_id: int) -> None:
...
+    for fd in reusable_form_definitions:
+        form_ids_to_update = fd.formstep_set.exclude(form=form).values_list(
+            "form", flat=True
+        )
+
+        for form_id in form_ids_to_update:
+            form_to_update = Form.objects.get(id=form_id)
+            form_to_update.formvariable_set.all().delete()
+
+            FormVariable.objects.create_for_form(form_to_update)
```
Comment (sergei-maertens) [inline src/openforms/forms/tasks.py:132]: I think you can simplify and optimize this:

```py
fds = FormDefinition.objects.filter(formstep__form=form_id, is_reusable=True)  # 0 queries
other_forms = Form.objects.filter(formstep__form_definition__in=fds).exclude(id=form_id)  # 0 queries
# delete the existing form variables, we will re-create them
FormVariable.objects.filter(form__in=other_forms, source=FormVariableSources.component).delete()  # 1 query
for form in other_forms:  # 2 queries
    FormVariable.objects.create_for_form(form)  # x number of queries for each form
```

We can avoid of nested loops (but you'll have to mark the task as `@transaction.atomic`!

The logic is:

1. Figure out which re-usable form definitions are involved
2. Get a collection of all the other forms that use any of these re-usable form definitions
3. Delete all their component form variables
4. Regenerate the component form variables for each form

File: src/openforms/forms/tasks.py [hunk 5]
```diff
@@ -87,6 +105,33 @@ def recouple_submission_variables_to_form_variables(form_id: int) -> None:
...
+    ).distinct()
+
+    for fd in reusable_form_definitions:
+        form_ids_to_update = fd.formstep_set.exclude(form=form).values_list(
+            "form", flat=True
+        )
+
+        for form_id in form_ids_to_update:
+            form_to_update = Form.objects.get(id=form_id)
+            form_to_update.formvariable_set.all().delete()
```
Comment (sergei-maertens) [inline src/openforms/forms/tasks.py:130]: this is also destroying the user defined variables, which is *not* okay. Please add a test case of a form with user defined variables too and assert that those are not broken.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/forms/tasks.py (hunk 5 above)
Comment (vaszig) [reply src/openforms/forms/tasks.py:130]: Yes, this was done by accident. At the beginning I was filtering out only `FormVariableSources.component` sources but yes, I will add a test for this as well.

**Review by sergei-maertens — APPROVED** [review]

**Comment by sergei-maertens** [comment]
Backports:

* 2.5.x: 0f5cdba087d549a82b0b00350d59d0abcf5ae19f
* 2.4.x: afafb954903c5247d262c4fb7d04be198af34041
* 2.3.x: ad3eabd3110f0343f7766f1f55c1b46e12c0facc

**Pull request merged**